sql/2021/security/csp_number_of_allowed_hosts.sql

```diff
@@ -0,0 +1,42 @@
+#standardSQL
+# CSP on home pages: number of unique headers, header length and number of allowed hosts in all directives
+CREATE TEMPORARY FUNCTION getHeader(headers STRING, headername STRING)
+RETURNS STRING DETERMINISTIC
+LANGUAGE js AS '''
+  const parsed_headers = JSON.parse(headers);
+  const matching_headers = parsed_headers.filter(h => h.name.toLowerCase() == headername.toLowerCase());
+  if (matching_headers.length > 0) {
+    return matching_headers[0].value;
+  }
+  return null;
+''';
+CREATE TEMP FUNCTION getNumUniqueHosts(str STRING) AS (
+  (SELECT COUNT(DISTINCT x) FROM UNNEST(REGEXP_EXTRACT_ALL(str, r'(?i)(https*://[^\s;]+)[\s;]')) AS x)
+);
+
+SELECT
+  client,
+  percentile,
+  COUNT(0) AS total_requests,
```

Comment:
This will not be the total number of requests because of the `UNNEST` statement.
See: https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Analysts'-Guide#a-warning-on-totals

Do you even need the total requests here? Normally don't include them in percentile queries.

Reply:
`UNNEST` doesn't appear to have an impact on the total number of requests here, since `percentile` is also part of the `GROUP BY` expression. So in other words: the `UNNEST` statement indeed multiplies the number of rows by 5, but since the resulting rows are also grouped by `percentile`, `COUNT(0)` only counts the number of requests per percentile.

Reply:
Yeah this was me being silly. All good!